File: kernel/api/src/main/java/ink/reactor/kernel/scheduler/Ticks.java
package ink.reactor.kernel.scheduler;

import java.time.Duration;

public record Ticks(int duration) {

    private static final Ticks NONE = new Ticks(0);

    public static Ticks none() {
        return NONE;
    }

    public static Ticks ofMillis(final int millis) {
        return new Ticks(TickUnit.MILLIS.toTicks(millis));
    }

    public static Ticks ofSeconds(final int seconds) {
        return new Ticks(TickUnit.SECONDS.toTicks(seconds));
    }

    public static Ticks ofMinutes(final int minutes) {
        return new Ticks(TickUnit.MINUTES.toTicks(minutes));
    }

    public static Ticks ofHours(final int hours) {
        return new Ticks(TickUnit.HOURS.toTicks(hours));
    }

    public static Ticks ofDays(final int days) {
        return new Ticks(TickUnit.DAYS.toTicks(days));
    }

    public static Ticks from(final Duration duration) {
        return new Ticks((int) (duration.toMillis() / TickUnit.MILLIS_PER_TICK));
    }
}

File: kernel/api/src/main/kotlin/ink/reactor/kernel/scheduler/Ticks.kt
package ink.reactor.kernel.scheduler

import java.time.Duration

@JvmInline
value class Ticks(val duration: Int) : Comparable<Ticks> {

    companion object {
        val NONE = Ticks(0)
        const val MILLIS_PER_TICK = 50L

        fun ofMillis(millis: Number) = Ticks((millis.toLong() / MILLIS_PER_TICK).toInt())
        fun ofSeconds(seconds: Number) = ofMillis(seconds.toLong() * 1000)
        fun ofMinutes(minutes: Number) = ofSeconds(minutes.toLong() * 60)
        fun ofHours(hours: Number) = ofMinutes(hours.toLong() * 60)
        fun ofDays(days: Number) = ofHours(days.toLong() * 24)
        fun ofYears(years: Number) = ofDays(years.toLong() * 365)

        fun from(duration: Duration) = ofMillis(duration.toMillis())
    }

    operator fun plus(other: Ticks) = Ticks(duration + other.duration)
    operator fun minus(other: Ticks) = Ticks(duration - other.duration)
    override fun compareTo(other: Ticks) = duration.compareTo(other.duration)
}
